experimental/dds/tree2/src/feature-libraries/sequence-field/compose.ts

```diff
@@ -666,182 +460,52 @@ export class ComposeQueue<T> {
-						// in the base changeset because we're in the process of merging in its inverse.
-						// If we had already encountered this base detach of content Y then we would have cancelled it
-						// out with (or ordered it with respect to) newMark.
-						// This later detach must therefore be present in the base changeset, and further to the right.
-						// We'll keep returning all the base marks before that.
-						return this.dequeueBase();
-					} else {
-						// The reattach is for a detach that occurred chronologically before the baseMark detach.
-						// We rely on the lineage information to tell us where in relation to baseMark this earlier
-						// detach was.
-						const currentOffset = this.baseGap.getOffset(baseRev);
-						const remainingOffset = targetOffset - currentOffset;
-						assert(remainingOffset >= 0, 0x4e0 /* Overshot the target gap */);
-						if (remainingOffset === 0) {
-							return this.dequeueNew();
-						}
-						return {
-							baseMark:
-								remainingOffset < getInputLength(baseMark)
-									? this.baseMarks.dequeueInput(remainingOffset)
-									: this.baseMarks.dequeue(),
-						};
-					}
-				}
-			} else if (
-				isReattach(newMark) &&
-				isReattach(baseMark) &&
-				areRelatedReattaches(baseMark, newMark)
-			) {
-				return dequeueRelatedReattaches(this.newMarks, this.baseMarks);
-			} else if (
-				newMark.type === "Insert" &&
-				newRev !== undefined &&
-				this.cancelledInserts.has(newRev)
-			) {
-				// We know the new insert is getting cancelled out so we need to delay returning it.
-				// The base mark that cancels the insert must appear later in the base marks.
-				return { baseMark: this.baseMarks.dequeue() };
-			} else if (
-				isDeleteMark(baseMark) &&
-				baseMark.revision !== undefined &&
-				this.revisionMetadata.getInfo(baseMark.revision).rollbackOf !== undefined
-			) {
-				// The base mark represents an insert being rolled back.
-				// That insert was concurrent to and sequenced after the attach performed by newNark so
-				// the delete should be ordered as the later insert would have been had the changes applied in
-				// sequencing order. This means the delete should come first since right now we only support
-				// the merge-left tiebreak policy.
-				// TODO: support merge-right tiebreak policy.
-				return { baseMark: this.baseMarks.dequeue() };
+				baseCellId = baseMark.detachEvent;
 			}
-			return this.dequeueNew();
-		} else if (isDetachMark(baseMark) || isBlockedReattach(baseMark)) {
+			const cmp = compareCellPositions(
+				baseCellId,
+				baseMark,
+				newMark,
+				this.newRevision,
+				this.cancelledInserts,
+				this.baseGap,
+			);
+			if (cmp < 0) {
+				return { baseMark: this.baseMarks.dequeueUpTo(-cmp) };
+			} else if (cmp > 0) {
+				return { newMark: this.newMarks.dequeueUpTo(cmp) };
+			} else {
+				const length = Math.min(getMarkLength(baseMark), getMarkLength(newMark));
+				return this.dequeueLength(length);
```

Comment:
Make `dequeueLength` automatically dequeue the `Math.min(getMarkLength(baseMark), getMarkLength(newMark))` by default since that's how it's used.
This also has the benefit making the contract clearer: `dequeueLength` will return two marks of the same length.

Reply:
Done